endpoints/event.go

```diff
@@ -0,0 +1,367 @@
+
+	// timestamp
+	if request.Timestamp > 0 {
+		r = r + nameValueAsQueryString(TIMESTAMP_PARAMETER, strconv.FormatInt(request.Timestamp, 10))
+	}
+
+	// bidder
+	if request.Bidder != "" {
+		r = r + nameValueAsQueryString(BIDDER_PARAMETER, request.Bidder)
+	}
+
+	// format
+	switch request.Format {
+	case BLANK:
+		r = r + nameValueAsQueryString(FORMAT_PARAMETER, string(BLANK))
+	case IMAGE:
+		r = r + nameValueAsQueryString(FORMAT_PARAMETER, string(IMAGE))
+	}
+
+	//analytics
+	switch request.Analytics {
+	case ENABLED:
+		r = r + nameValueAsQueryString(ANALYTICS_PARAMETER, string(ENABLED))
+	case DISABLED:
+		r = r + nameValueAsQueryString(ANALYTICS_PARAMETER, string(DISABLED))
+	}
+
+	return r
+}
+
+func nameValueAsQueryString(name string, value string) string {
+	if value == "" {
+		return ""
+	}
+
+	return "&" + name + "=" + value
+}
+
+/**
+ * validate type
+ */
+func validateType(er *EventRequest, httpRequest *http.Request) error {
+	t, err := validateRequiredParameter(httpRequest, TYPE_PARAMETER)
+
+	if err != nil {
+		return err
+	}
+
+	switch t {
+	case string(IMP):
+		er.Type = IMP
+		return nil
+	case string(WIN):
+		er.Type = WIN
+		return nil
+	default:
+		return fmt.Errorf("unknown type: '%s'", t)
+	}
+}
+
+/**
+ * validate format
+ */
+func validateFormat(er *EventRequest, httpRequest *http.Request) error {
+	f := httpRequest.FormValue(FORMAT_PARAMETER)
+
+	if f != "" {
+		switch f {
+		case string(BLANK):
+			er.Format = BLANK
+			return nil
+		case string(IMAGE):
+			er.Format = IMAGE
+			return nil
+		default:
+			return fmt.Errorf("unknown format: '%s'", f)
+		}
+	}
+
+	return nil
+}
+
+/**
+ * validate analytics
+ */
+func validateAnalytics(er *EventRequest, httpRequest *http.Request) error {
+	a := httpRequest.FormValue(ANALYTICS_PARAMETER)
+
+	if a != "" {
+		switch a {
+		case string(ENABLED):
+			er.Analytics = ENABLED
+			return nil
+		case string(DISABLED):
+			er.Analytics = DISABLED
+			return nil
+		default:
+			return fmt.Errorf("unknown analytics: '%s'", a)
+		}
+	}
+
+	return nil
+}
+
+/**
+ * validate timestamp
+ */
+func validateTimestamp(er *EventRequest, httpRequest *http.Request) error {
+	t := httpRequest.FormValue(TIMESTAMP_PARAMETER)
+
+	if t != "" {
+		ts, err := strconv.ParseInt(t, 10, 64)
+
+		if err != nil {
+			return fmt.Errorf("invalid request: error parsing timestamp '%s'", t)
+		}
+
+		er.Timestamp = ts
+		return nil
+	}
+
+	return nil
+}
+
+/**
+ * validate required parameter
+ */
+func validateRequiredParameter(httpRequest *http.Request, parameter string) (string, error) {
+	t := httpRequest.FormValue(parameter)
+
+	if t == "" {
+		return "", fmt.Errorf("parameter '%s' is required", parameter)
+	}
+
+	return t, nil
+}
+
+/**
+ * Load tracking pixel from Base64 string
+ */
+func loadTrackingPixel() *TrackingPixel {
+	reader := base64.NewDecoder(base64.StdEncoding, strings.NewReader(TRACKING_PIXEL_PNG))
+	buff := bytes.Buffer{}
+	_, err := buff.ReadFrom(reader)
+	if err != nil {
+		panic(err)
```

Comment:
A panic seems ok here since it's decoding a constant. An alternative is to encode the byte array directly:

```
[]byte{0x00, 0x00, ..}
```

It's a pain, for sure.. but a one time pain. :)